File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        
        def check(string):
            return string == string[::-1]
        
        n = len(s)
        answer = []
        stack = []
        def backtrack(index):
            if index == n:
                answer.append(stack[:])
            
            for i in range(index + 1, n + 1):
                subString = s[index : i]
                if not check(subString):
                    continue
                stack.append(subString)
                backtrack(i)
                stack.pop()
        
        backtrack(0)
        return answer

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
    
    function isPalindrome(s){
        for (let i = 0; i < s.length/2; i++){
            if (s[i] != s[s.length - i  - 1]) return false
        }
        
        return true
    }
    
    function dp(index){
        if (index == -1) return [[]]
        const possible = []
        for (let i = index; i >= 0; i--){
            const substring = s.substring(i, index + 1)
            if (!isPalindrome(substring)) continue
            dp(i - 1).forEach(element => {
                const newElement = [...element, substring]
                possible.push(newElement)
            })
        }
        return possible
    }
    return dp(s.length - 1)
};